File: src/session/lock.ts
import fs from "node:fs";
import path from "node:path";

import { Type } from "@sinclair/typebox";
import type { Static } from "@sinclair/typebox";

import { assertNever, GymratError, hasErrorCode, messageOf } from "../errors.js";
import { compile, expected } from "../schema.js";

const lockHolderSchema = Type.Object(
  {
    // The compiled check requires a finite number, so a `pid` that overflowed
    // JSON's number range reads as a foreign shape rather than as a holder no
    // signal can ever reach.
    pid: Type.Number(expected("a number")),
    command: Type.String(expected("a string")),
    at: Type.String(expected("a string")),
  },
  // Deliberately not strict about unknown keys: a lockfile carrying a field
  // this version does not know is still a live holder, and rejecting it would
  // steal the lock out from under whoever wrote it.
  expected("an object"),
);

/** The process a lockfile records as its holder. */
type LockHolder = Static<typeof lockHolderSchema>;

const lockHolderValidator = compile(lockHolderSchema);

/**
 * Gives up an acquired lock. Calling it more than once is harmless.
 *
 * Only the lockfile this run published is removed. A lock taken over since
 * belongs to whoever holds it now, and is left where it stands.
 */
export type ReleaseLock = () => void;

/** How many times acquisition re-reads a lockfile it lost a race for. */
const MAX_ACQUIRE_ATTEMPTS = 3;

/**
 * Which file a lockfile read came from, as the filesystem identifies it.
 *
 * A path says nothing about which file it names from one moment to the next, so
 * a steal that only knows `lockPath` cannot tell the record it judged stale from
 * whatever a rival published there since. Device and inode do tell them apart.
 */
type LockfileIdentity = {
  readonly dev: bigint;
  readonly ino: bigint;
};

/** Whether two reads came from one and the same file. */
function sameFile(one: LockfileIdentity, other: LockfileIdentity): boolean {
  return one.dev === other.dev && one.ino === other.ino;
}

/** Which file an open descriptor refers to, whatever its path names by now. */
function identityOf(fd: number): LockfileIdentity {
  const stats = fs.fstatSync(fd, { bigint: true });
  return { dev: stats.dev, ino: stats.ino };
}

/** What a lockfile says at the moment it was read. */
type LockfileState =
  | { readonly kind: "absent" }
  | { readonly kind: "held"; readonly holder: LockHolder; readonly identity: LockfileIdentity }
  | { readonly kind: "unreadable"; readonly identity: LockfileIdentity };

/**
 * Read what the lockfile at `lockPath` currently says, and which file said it.
 *
 * The identity is taken from the open descriptor the contents are read through,
 * so the two describe one file even when the lock path is taken over mid-read.
 *
 * A file that cannot be parsed, or that carries a foreign shape, is reported as
 * `unreadable`. Publication is atomic, so no reader ever catches a holder
 * mid-write: an unreadable file is debris — a run killed between writing its
 * record and publishing it, or a foreign file at the lock path — not a lock
 * somebody is in the middle of taking.
 *
 * @throws GymratError when the lockfile belongs to another user.
 */
function readLockfile(lockPath: string): LockfileState {
  let fd: number;
  try {
    fd = fs.openSync(lockPath, "r");
  } catch (error) {
    if (hasErrorCode(error, "ENOENT")) {
      return { kind: "absent" };
    }
    // A lockfile another user owns is unreadable to every later run, so the
    // steal path below can never reach it — the only way out is by hand.
    if (hasErrorCode(error, "EPERM") || hasErrorCode(error, "EACCES")) {
      throw new GymratError(
        `Lock file ${lockPath} could not be read: ${messageOf(error)}`,
        `It belongs to another user. Remove ${lockPath} yourself, then rerun.`,
        { cause: error },
      );
    }
    throw error;
  }

  let identity: LockfileIdentity;
  let contents: string;
  try {
    identity = identityOf(fd);
    contents = fs.readFileSync(fd, "utf8");
  } finally {
    fs.closeSync(fd);
  }

  let parsed: unknown;
  try {
    parsed = JSON.parse(contents);
  } catch {
    return { kind: "unreadable", identity };
  }
  return lockHolderValidator.check(parsed)
    ? { kind: "held", holder: parsed, identity }
    : { kind: "unreadable", identity };
}

/**
 * Whether a process with `pid` still exists.
 *
 * Signal `0` runs the kernel's permission and existence checks without
 * delivering anything. Only `ESRCH` means no such process: `EPERM` says the
 * process is there but owned by another user, which is still a live holder.
 */
function isAlive(pid: number): boolean {
  try {
    process.kill(pid, 0);
    return true;
  } catch (error) {
    return !hasErrorCode(error, "ESRCH");
  }
}

/** Delete `filePath` if it exists. Used on the lockfile and its scratch siblings alike. */
function unlinkIfExists(filePath: string): void {
  try {
    fs.unlinkSync(filePath);
  } catch (error) {
    if (!hasErrorCode(error, "ENOENT")) {
      throw error;
    }
  }
}

/**
 * Delete `lockPath` only while it still names the file `identity` came from.
 *
 * A path is not a lock: the file a run published can be displaced by a takeover
 * at any moment, and deleting whatever answers to the path would hand the next
 * holder's lock away to nobody. A different file there — or none at all — is
 * left exactly as found.
 */
function unlinkIfSameFile(lockPath: string, identity: LockfileIdentity): void {
  let fd: number;
  try {
    fd = fs.openSync(lockPath, "r");
  } catch (error) {
    if (hasErrorCode(error, "ENOENT")) {
      return;
    }
    throw error;
  }

  let current: LockfileIdentity;
  try {
    current = identityOf(fd);
  } finally {
    fs.closeSync(fd);
  }

  if (sameFile(current, identity)) {
    unlinkIfExists(lockPath);
  }
}

/**
 * Report a lock whose holder answered a liveness probe.
 *
 * Deleting the lockfile is deliberately not offered: the holder is provably
 * alive, so the only safe remedy is to wait it out.
 */
function heldByError(holder: LockHolder): GymratError {
  return new GymratError(
    `Lock held by PID ${String(holder.pid)} (${holder.command}, started ${holder.at})`,
    "Another gymrat run is active in this repo. Wait for it to finish.",
  );
}

/**
 * Publish `record` at `lockPath`, handing back the identity of the file now
 * standing there, or `undefined` when someone else got there first.
 *
 * The record is written to a scratch file beside the lock and only then linked
 * into place, so the lock path never exists holding half a record: readers see a
 * whole holder or nothing at all. The link is also the exclusive step — it fails
 * with `EEXIST` when the path is taken — so exactly one racer publishes.
 *
 * The identity is taken from the descriptor the record was written through,
 * which is the same file the link names, so it stays true no matter what takes
 * the lock path over afterwards.
 */
function publishLockRecord(lockPath: string, record: string): LockfileIdentity | undefined {
  const scratchPath = `${lockPath}.${String(process.pid)}.record`;
  const fd = fs.openSync(scratchPath, "w");
  let identity: LockfileIdentity;
  try {
    fs.writeFileSync(fd, record);
    identity = identityOf(fd);
  } finally {
    fs.closeSync(fd);
  }

  try {
    fs.linkSync(scratchPath, lockPath);
    return identity;
  } catch (error) {
    if (hasErrorCode(error, "EEXIST")) {
      return undefined;
    }
    throw error;
  } finally {
    unlinkIfExists(scratchPath);
  }
}

/**
 * Rethrow the failure to displace the stale lockfile at `lockPath`.
 *
 * `EPERM` and `EACCES` say the file belongs to another user — a sticky `/tmp` —
 * which no retry can resolve, so they are framed for whoever has to clean up.
 */
function rethrowDisplacementFailure(lockPath: string, error: unknown): never {
  if (hasErrorCode(error, "EPERM") || hasErrorCode(error, "EACCES")) {
    throw new GymratError(
      `Stale lock file ${lockPath} could not be removed: ${messageOf(error)}`,
      `It belongs to another user. Remove ${lockPath} yourself, then rerun.`,
      { cause: error },
    );
  }
  throw error;
}

/**
 * What asking for the right to displace a stale lockfile turned up.
 *
 * `gone` covers both ways the judged file can stop being the one at the lock
 * path: it vanished before the claim, or the claim came back naming a different
 * file.
 */
type ClaimOutcome =
  | { readonly kind: "claimed"; readonly claimPath: string }
  | { readonly kind: "blocked"; readonly claimPath: string }
  | { readonly kind: "gone" };

/**
 * Claim the right to displace the file `identity` was read from, handing back
 * the path the claim lives at.
 *
 * The claim is a second name for the lockfile itself, spelled out of that file's
 * identity, which makes it both halves of a safe steal. It is exclusive: two
 * racers reaching the same staleness verdict about one file ask for the same
 * name, `EEXIST` tells the loser so, and only the winner goes on to displace
 * anything. And it is proof: a racer whose claim turns out to name a different
 * file is holding a lock somebody published after the verdict, so it drops the
 * claim and leaves that lock where it stands.
 *
 * A steal that is not this racer's to make is reported as `blocked` — the claim
 * name is taken — or as `gone`, leaving acquisition to read the lock path
 * afresh. The two are told apart because a claim that stays taken across every
 * attempt is the fingerprint of a run that died holding it.
 *
 * @throws GymratError when the lockfile belongs to another user.
 */
function claimStaleLock(lockPath: string, identity: LockfileIdentity): ClaimOutcome {
  const claimPath = `${lockPath}.${String(identity.dev)}-${String(identity.ino)}.claim`;
  try {
    fs.linkSync(lockPath, claimPath);
  } catch (error) {
    if (hasErrorCode(error, "EEXIST")) {
      return { kind: "blocked", claimPath };
    }
    if (hasErrorCode(error, "ENOENT")) {
      return { kind: "gone" };
    }
    rethrowDisplacementFailure(lockPath, error);
  }

  const claimed = fs.statSync(claimPath, { bigint: true });
  if (sameFile(claimed, identity)) {
    return { kind: "claimed", claimPath };
  }
  unlinkIfExists(claimPath);
  return { kind: "gone" };
}

/**
 * Clear the claimed lockfile off the lock path. Returns whether it went away.
 *
 * Only the holder of the claim gets here, and nothing else may displace a
 * claimed file, so the file moved is the one that was judged stale. It is moved
 * rather than deleted so a run killed mid-steal leaves the lock path free rather
 * than holding a record it never published.
 *
 * @throws GymratError when the lockfile belongs to another user.
 */
function displaceStaleLock(lockPath: string): boolean {
  const asidePath = `${lockPath}.${String(process.pid)}.stale`;
  try {
    fs.renameSync(lockPath, asidePath);
  } catch (error) {
    if (hasErrorCode(error, "ENOENT")) {
      return false;
    }
    rethrowDisplacementFailure(lockPath, error);
  }
  unlinkIfExists(asidePath);
  return true;
}

/**
 * What taking over a lockfile no live process holds turned up.
 *
 * `blocked` carries the claim path that stood in the way, because a claim
 * nobody is behind can only be cleared by hand.
 */
type StealOutcome =
  | { readonly kind: "won"; readonly identity: LockfileIdentity }
  | { readonly kind: "lost" }
  | { readonly kind: "blocked"; readonly claimPath: string };

/** Take over a lockfile no live process holds. */
function stealLock(lockPath: string, identity: LockfileIdentity, record: string): StealOutcome {
  const claim = claimStaleLock(lockPath, identity);
  switch (claim.kind) {
    case "gone":
      return { kind: "lost" };
    case "blocked":
      return { kind: "blocked", claimPath: claim.claimPath };
    case "claimed":
      try {
        const published = displaceStaleLock(lockPath)
          ? publishLockRecord(lockPath, record)
          : undefined;
        return published === undefined ? { kind: "lost" } : { kind: "won", identity: published };
      } finally {
        unlinkIfExists(claim.claimPath);
      }
    default:
      return assertNever(claim);
  }
}

/**
 * A takeover that died holding its claim, leaving the lock impossible to steal.
 *
 * The claim outlives the run that made it, so every later steal of that same
 * file is refused the claim name forever. `holderPid` is the dead process the
 * lockfile named, or `undefined` when the lockfile was too damaged to name one.
 */
type WedgedTakeover = {
  readonly identity: LockfileIdentity;
  readonly claimPath: string;
  readonly holderPid: number | undefined;
};

function wedgedTakeoverError(lockPath: string, wedge: WedgedTakeover): GymratError {
  const nobodyHolds =
    wedge.holderPid === undefined
      ? "No gymrat process holds this lock."
      : `No gymrat process holds this lock (PID ${String(wedge.holderPid)} is dead).`;
  return new GymratError(
    `Lock at ${lockPath} was left behind by a run that died while taking it over.`,
    `${nobodyHolds} To unblock, delete ${lockPath} and ${wedge.claimPath}, then rerun.`,
  );
}

/** What one pass at the lock path turned up. */
type AttemptOutcome =
  | { readonly kind: "acquired"; readonly identity: LockfileIdentity }
  | { readonly kind: "retry" }
  | { readonly kind: "blocked"; readonly wedge: WedgedTakeover };

/** Read a steal that did not win as either worth retrying or as wedged. */
function attemptFromSteal(
  steal: StealOutcome,
  identity: LockfileIdentity,
  holderPid: number | undefined,
): AttemptOutcome {
  switch (steal.kind) {
    case "won":
      return { kind: "acquired", identity: steal.identity };
    case "lost":
      return { kind: "retry" };
    case "blocked":
      return { kind: "blocked", wedge: { identity, claimPath: steal.claimPath, holderPid } };
    default:
      return assertNever(steal);
  }
}

/**
 * Make one bid for the lock at `lockPath`: publish, or judge what is there.
 *
 * @throws GymratError when the lock is held by a process that is still running,
 *   or when the lockfile belongs to another user.
 */
function attemptAcquire(lockPath: string, record: string): AttemptOutcome {
  const published = publishLockRecord(lockPath, record);
  if (published !== undefined) {
    return { kind: "acquired", identity: published };
  }

  const state = readLockfile(lockPath);
  switch (state.kind) {
    case "absent":
      return { kind: "retry" };
    case "unreadable":
      return attemptFromSteal(
        stealLock(lockPath, state.identity, record),
        state.identity,
        undefined,
      );
    case "held": {
      if (isAlive(state.holder.pid)) {
        throw heldByError(state.holder);
      }
      const steal = stealLock(lockPath, state.identity, record);
      return attemptFromSteal(steal, state.identity, state.holder.pid);
    }
    default:
      return assertNever(state);
  }
}

/**
 * Take the single-flight lock at `lockPath` on behalf of `command`.
 *
 * The lockfile is published exclusively, so two processes racing for it cannot
 * both win. A lockfile no live process holds is stolen silently — a crashed run
 * must not need manual cleanup, whether it left a holder record behind or a file
 * too damaged to read — and losing that steal re-enters acquisition, where the
 * winner is either a live holder to report or a lock released again in the
 * meantime.
 *
 * One exception: a run killed between claiming the right to displace a stale
 * lockfile and completing that displacement leaves a state no later run can
 * clear on its own. The thrown error names both files to delete.
 *
 * @throws GymratError when the lock is held by a process that is still running,
 *   when the lockfile belongs to another user, or when every attempt was refused
 *   by the claim of a takeover that never finished.
 */
export function acquireLock(lockPath: string, command: string): ReleaseLock {
  const holder: LockHolder = { pid: process.pid, command, at: new Date().toISOString() };
  const record = JSON.stringify(holder);

  fs.mkdirSync(path.dirname(lockPath), { recursive: true });

  let wedge: WedgedTakeover | undefined;
  let wedgedEveryAttempt = true;

  for (let attempt = 0; attempt < MAX_ACQUIRE_ATTEMPTS; attempt++) {
    const outcome = attemptAcquire(lockPath, record);
    if (outcome.kind === "acquired") {
      const { identity } = outcome;
      return () => {
        unlinkIfSameFile(lockPath, identity);
      };
    }

    // Only one file blocking every single attempt rules out the rival that
    // takes the lock, works, and releases it between two of our reads.
    const blocked = outcome.kind === "blocked" ? outcome.wedge : undefined;
    if (blocked === undefined) {
      wedgedEveryAttempt = false;
    } else if (wedge === undefined) {
      wedge = blocked;
    } else if (!sameFile(wedge.identity, blocked.identity)) {
      wedgedEveryAttempt = false;
    }
  }

  if (wedgedEveryAttempt && wedge !== undefined) {
    throw wedgedTakeoverError(lockPath, wedge);
  }

  throw new GymratError(
    `Lock at ${lockPath} was claimed by another process on every attempt.`,
    `Another gymrat run is active in this repo. Wait for it to finish. ` +
      `If no gymrat process is running, delete ${lockPath}.`,
  );
}

File: src/gymrat_py/session/lock.py
"""Single-flight repository lock built on the atomicity of ``os.link``.

A run publishes its holder record by writing it to a scratch file and hard-linking
that file into place: the link is exclusive — it fails with ``EEXIST`` when the
path is taken — so exactly one racer wins. A lockfile no live process holds is
stolen silently through a claim link derived from the file's device/inode, so a
crashed run never needs manual cleanup. The one state no later run can clear on
its own — a takeover killed between claiming and completing a displacement — is
reported with both files to delete.
"""

import errno
import json
import os
import sys
from collections.abc import Callable
from dataclasses import dataclass
from pathlib import Path
from typing import NoReturn

from gymrat_py.errors import GymratError, message_of
from gymrat_py.session.clock import now_iso

# Largest pid a liveness probe can be asked about: signalling rejects anything
# past int32, and 0/negative address process groups rather than a process.
MAX_PID = 2_147_483_647

# How many times acquisition re-reads a lockfile it lost a race for.
MAX_ACQUIRE_ATTEMPTS = 3

# The remedy for a lock whose holder answered a liveness probe. Deleting the
# lockfile is deliberately not offered: the holder is provably alive.
_LIVE_HOLDER_HINT = "Another gymrat run is active in this repo. Wait for it to finish."


type ReleaseLock = Callable[[], None]
"""Gives up an acquired lock. Calling it more than once is harmless.

Only the lockfile this run published is removed. A lock taken over since belongs
to whoever holds it now, and is left where it stands.
"""


@dataclass(frozen=True)
class LockHolder:
    """The process a lockfile records as its holder."""

    pid: int
    command: str
    at: str


@dataclass(frozen=True)
class LockIdentity:
    """Which file a lockfile read came from, as the filesystem identifies it.

    A path says nothing about which file it names from one moment to the next, so
    a steal that only knows the lock path cannot tell the record it judged stale
    from whatever a rival published there since. Device and inode do tell them
    apart.
    """

    dev: int
    ino: int


def _parse_holder(value: object) -> LockHolder | None:
    """Return the holder ``value`` describes, or ``None`` when it is debris.

    A pid outside ``[1, MAX_PID]`` — a fraction, 0, a negative, or a number past
    int32 — names no process any signal could reach, so the record is debris.
    Unknown extra keys are tolerated: a lockfile carrying a field this version
    does not know is still a live holder, and rejecting it would steal the lock
    out from under whoever wrote it.
    """
    if not isinstance(value, dict):
        return None
    pid = value.get("pid")
    command = value.get("command")
    at = value.get("at")
    if isinstance(pid, bool) or not isinstance(pid, int) or pid < 1 or pid > MAX_PID:
        return None
    if not isinstance(command, str) or not isinstance(at, str):
        return None
    return LockHolder(pid=pid, command=command, at=at)


def _identity_of(fd: int) -> LockIdentity:
    """Which file an open descriptor refers to, whatever its path names by now."""
    info = os.fstat(fd)
    return LockIdentity(dev=info.st_dev, ino=info.st_ino)


def _read_all(fd: int) -> str:
    """Read the whole contents of ``fd`` as UTF-8 text."""
    chunks: list[bytes] = []
    while chunk := os.read(fd, 65536):
        chunks.append(chunk)
    return b"".join(chunks).decode("utf-8")


def _write_all(fd: int, data: bytes) -> None:
    """Write every byte of ``data`` to ``fd``."""
    view = memoryview(data)
    while view:
        view = view[os.write(fd, view) :]


def _force_unlink(path: str) -> None:
    """Remove ``path``, treating an already-absent file as success."""
    try:
        os.unlink(path)  # noqa: PTH108
    except FileNotFoundError:
        return


# What a lockfile says at the moment it was read.
@dataclass(frozen=True)
class _Absent:
    """No file stands at the lock path."""


@dataclass(frozen=True)
class _Held:
    """A parseable holder record and the file it was read from."""

    holder: LockHolder
    identity: LockIdentity


@dataclass(frozen=True)
class _Unreadable:
    """Debris at the lock path — unparseable or a foreign shape — and its file."""

    identity: LockIdentity


type _LockfileState = _Absent | _Held | _Unreadable


def read_lockfile(lock_path: str) -> _LockfileState:
    """Read what the lockfile at ``lock_path`` currently says, and which file said it.

    The identity is taken from the open descriptor the contents are read through,
    so the two describe one file even when the lock path is taken over mid-read.
    Publication is atomic, so no reader ever catches a holder mid-write: an
    unreadable file is debris — a run killed before publishing, or a foreign file
    at the lock path — not a lock somebody is in the middle of taking.

    Raises:
        GymratError: When the lockfile belongs to another user.
    """
    try:
        fd = os.open(lock_path, os.O_RDONLY)
    except FileNotFoundError:
        return _Absent()
    except PermissionError as error:
        # A lockfile another user owns is unreadable to every later run, so the
        # steal path can never reach it — the only way out is by hand.
        message = f"Lock file {lock_path} could not be read: {message_of(error)}"
        hint = f"It belongs to another user. Remove {lock_path} yourself, then rerun."
        raise GymratError(message, hint=hint) from error

    try:
        identity = _identity_of(fd)
        contents = _read_all(fd)
    finally:
        os.close(fd)

    try:
        parsed = json.loads(contents)
    except ValueError:
        return _Unreadable(identity)
    holder = _parse_holder(parsed)
    return _Held(holder, identity) if holder is not None else _Unreadable(identity)


def is_alive(pid: int) -> bool:
    """Whether a process with ``pid`` still exists.

    Signal ``0`` runs the kernel's permission and existence checks without
    delivering anything. Only ``ESRCH`` means no such process: ``EPERM`` says the
    process is there but owned by another user, which is still a live holder.
    """
    try:
        os.kill(pid, 0)
    except OSError as error:
        return error.errno != errno.ESRCH
    return True


def unlink_if_same_file(lock_path: str, identity: LockIdentity) -> None:
    """Delete ``lock_path`` only while it still names the file ``identity`` came from.

    A path is not a lock: the file a run published can be displaced by a takeover
    at any moment, and deleting whatever answers to the path would hand the next
    holder's lock away to nobody. A different file there — or none at all — is
    left exactly as found.
    """
    try:
        fd = os.open(lock_path, os.O_RDONLY)
    except FileNotFoundError:
        return

    try:
        current = _identity_of(fd)
    finally:
        os.close(fd)

    if current == identity:
        _force_unlink(lock_path)


def still_names_file(lock_path: str, identity: LockIdentity) -> bool:
    """Whether ``lock_path`` still names the file ``identity`` came from.

    The answer is a snapshot, true only for the instant it was taken — enough to
    word a remedy about the file standing at the path, never enough to decide a
    write against it. :func:`unlink_if_same_file` is the safe form for that.
    """
    try:
        info = os.stat(lock_path)  # noqa: PTH116
    except FileNotFoundError:
        return False
    return LockIdentity(dev=info.st_dev, ino=info.st_ino) == identity


def _held_by_error(holder: LockHolder) -> GymratError:
    """Report a lock whose holder answered a liveness probe."""
    message = f"Lock held by PID {holder.pid} ({holder.command}, started {holder.at})"
    return GymratError(message, hint=_LIVE_HOLDER_HINT)


def publish_lock_record(lock_path: str, record: str) -> LockIdentity | None:
    """Publish ``record`` at ``lock_path``, or ``None`` when someone got there first.

    The record is written to a scratch file beside the lock and only then linked
    into place, so the lock path never exists holding half a record: readers see a
    whole holder or nothing at all. The link is also the exclusive step — it fails
    with ``EEXIST`` when the path is taken — so exactly one racer publishes.

    The identity is taken from the descriptor the record was written through,
    which is the same file the link names, so it stays true no matter what takes
    the lock path over afterwards.
    """
    scratch_path = f"{lock_path}.{os.getpid()}.record"
    try:
        fd = os.open(scratch_path, os.O_WRONLY | os.O_CREAT | os.O_TRUNC, 0o644)
    except OSError as error:
        # A leftover another user owns at the scratch path — a sticky ``/tmp`` —
        # is reframed like every other auxiliary displacement wall, not thrown raw.
        _rethrow_displacement_failure(lock_path, error)
    try:
        _write_all(fd, record.encode("utf-8"))
        identity = _identity_of(fd)
    finally:
        os.close(fd)

    try:
        os.link(scratch_path, lock_path)
    except FileExistsError:
        return None
    else:
        return identity
    finally:
        _force_unlink(scratch_path)


def _rethrow_displacement_failure(lock_path: str, error: OSError) -> NoReturn:
    """Reframe a permission failure to displace a stale lockfile, else re-raise.

    ``EPERM`` and ``EACCES`` say the file belongs to another user — a sticky
    ``/tmp`` — which no retry can resolve, so they are framed for whoever has to
    clean up.
    """
    if isinstance(error, PermissionError):
        message = f"Stale lock file {lock_path} could not be removed: {message_of(error)}"
        hint = f"It belongs to another user. Remove {lock_path} yourself, then rerun."
        raise GymratError(message, hint=hint) from error
    raise error


# What asking for the right to displace a stale lockfile turned up. ``_ClaimGone``
# covers both ways the judged file can stop being the one at the lock path: it
# vanished before the claim, or the claim came back naming a different file.
@dataclass(frozen=True)
class _ClaimClaimed:
    """The judged file was claimed for displacement; its claim lives here."""

    claim_path: str


@dataclass(frozen=True)
class _ClaimBlocked:
    """The claim name is already taken; the blocking claim lives here."""

    claim_path: str


@dataclass(frozen=True)
class _ClaimGone:
    """The judged file is no longer the one at the lock path."""


type _ClaimOutcome = _ClaimClaimed | _ClaimBlocked | _ClaimGone


def claim_stale_lock(lock_path: str, identity: LockIdentity) -> _ClaimOutcome:
    """Claim the right to displace the file ``identity`` was read from.

    The claim is a second name for the lockfile itself, spelled out of that file's
    identity, which makes it both halves of a safe steal. It is exclusive: two
    racers reaching the same staleness verdict about one file ask for the same
    name, ``EEXIST`` tells the loser so. And it is proof: a racer whose claim
    turns out to name a different file is holding a lock somebody published after
    the verdict, so it drops the claim and leaves that lock where it stands.

    Raises:
        GymratError: When the lockfile belongs to another user.
    """
    claim_path = f"{lock_path}.{identity.dev}-{identity.ino}.claim"
    try:
        os.link(lock_path, claim_path)
    except FileExistsError:
        return _ClaimBlocked(claim_path)
    except FileNotFoundError:
        return _ClaimGone()
    except OSError as error:
        _rethrow_displacement_failure(lock_path, error)

    info = os.stat(claim_path)  # noqa: PTH116
    if LockIdentity(dev=info.st_dev, ino=info.st_ino) == identity:
        return _ClaimClaimed(claim_path)
    _force_unlink(claim_path)
    return _ClaimGone()


def displace_stale_lock(lock_path: str) -> bool:
    """Clear the claimed lockfile off the lock path. Returns whether it went away.

    Only the holder of the claim gets here, and nothing else may displace a
    claimed file, so the file moved is the one that was judged stale. It is moved
    rather than deleted so a run killed mid-steal leaves the lock path free rather
    than holding a record it never published.

    Raises:
        GymratError: When the lockfile belongs to another user.
    """
    aside_path = f"{lock_path}.{os.getpid()}.stale"
    try:
        # Raw os.rename is the atomic displacement seam the retry loop races on.
        os.rename(lock_path, aside_path)  # noqa: PTH104
    except FileNotFoundError:
        return False
    except OSError as error:
        _rethrow_displacement_failure(lock_path, error)
    _force_unlink(aside_path)
    return True


@dataclass(frozen=True)
class _WedgedTakeover:
    """A takeover that died holding its claim, leaving the lock impossible to steal.

    The claim outlives the run that made it, so every later steal of that same
    file is refused the claim name forever. ``holder_pid`` is the dead process the
    lockfile named, or ``None`` when the lockfile was too damaged to name one.
    """

    identity: LockIdentity
    claim_path: str
    holder_pid: int | None


# What one pass at the lock path turned up.
@dataclass(frozen=True)
class _Acquired:
    """The lock was published; the release handle keys on this identity."""

    identity: LockIdentity


@dataclass(frozen=True)
class _Retry:
    """The lock path shifted under the attempt; read it afresh."""


@dataclass(frozen=True)
class _Blocked:
    """A claim stood in the way of a steal."""

    wedge: _WedgedTakeover


type _AttemptOutcome = _Acquired | _Retry | _Blocked


def steal_lock(
    lock_path: str,
    identity: LockIdentity,
    record: str,
    holder_pid: int | None,
) -> _AttemptOutcome:
    """Take over a lockfile no live process holds."""
    claim = claim_stale_lock(lock_path, identity)
    if isinstance(claim, _ClaimGone):
        return _Retry()
    if isinstance(claim, _ClaimBlocked):
        return _Blocked(_WedgedTakeover(identity, claim.claim_path, holder_pid))
    try:
        published = (
            publish_lock_record(lock_path, record) if displace_stale_lock(lock_path) else None
        )
        return _Acquired(published) if published is not None else _Retry()
    finally:
        _force_unlink(claim.claim_path)


def _wedged_takeover_error(lock_path: str, wedge: _WedgedTakeover) -> GymratError:
    """Word the remedy for a lock a dead takeover left impossible to steal."""
    if wedge.holder_pid is None:
        nobody_holds = "No gymrat process holds this lock."
    else:
        nobody_holds = f"No gymrat process holds this lock (PID {wedge.holder_pid} is dead)."
    message = f"Lock at {lock_path} was left behind by a run that died while taking it over."
    hint = f"{nobody_holds} To unblock, delete {lock_path} and {wedge.claim_path}, then rerun."
    return GymratError(message, hint=hint)


def attempt_acquire(lock_path: str, record: str) -> _AttemptOutcome:
    """Make one bid for the lock at ``lock_path``: publish, or judge what is there.

    Raises:
        GymratError: When the lock is held by a process that is still running, or
            when the lockfile belongs to another user.
    """
    published = publish_lock_record(lock_path, record)
    if published is not None:
        return _Acquired(published)

    state = read_lockfile(lock_path)
    if isinstance(state, _Absent):
        return _Retry()
    if isinstance(state, _Unreadable):
        return steal_lock(lock_path, state.identity, record, None)
    if is_alive(state.holder.pid):
        raise _held_by_error(state.holder)
    return steal_lock(lock_path, state.identity, record, state.holder.pid)


def _make_release(lock_path: str, identity: LockIdentity) -> ReleaseLock:
    """Build a release handle that removes only the file this run published."""

    def release() -> None:
        try:
            unlink_if_same_file(lock_path, identity)
        except OSError as error:
            text = f"Warning: failed to release lock at {lock_path}: {message_of(error)}\n"
            sys.stderr.write(text)

    return release


def acquire_lock(lock_path: str, command: str) -> ReleaseLock:
    """Take the single-flight lock at ``lock_path`` on behalf of ``command``.

    The lockfile is published exclusively, so two processes racing for it cannot
    both win. A lockfile no live process holds is stolen silently — a crashed run
    must not need manual cleanup, whether it left a holder record behind or a file
    too damaged to read — and losing that steal re-enters acquisition, where the
    winner is either a live holder to report or a lock released again in the
    meantime.

    One exception: a run killed between claiming the right to displace a stale
    lockfile and completing that displacement leaves a state no later run can
    clear on its own. The thrown error names both files to delete.

    Raises:
        GymratError: When the lock is held by a process that is still running,
            when the lockfile belongs to another user, or when every attempt was
            refused by the claim of a takeover that never finished.
    """
    pid = os.getpid()
    record = json.dumps({"pid": pid, "command": command, "at": now_iso()})

    Path(lock_path).parent.mkdir(parents=True, exist_ok=True)

    wedge: _WedgedTakeover | None = None
    wedged_every_attempt = True

    for _ in range(MAX_ACQUIRE_ATTEMPTS):
        outcome = attempt_acquire(lock_path, record)
        if isinstance(outcome, _Acquired):
            return _make_release(lock_path, outcome.identity)

        # Only one file blocking every single attempt rules out the rival that
        # takes the lock, works, and releases it between two of our reads.
        blocked = outcome.wedge if isinstance(outcome, _Blocked) else None
        if blocked is None:
            wedged_every_attempt = False
        elif wedge is None:
            wedge = blocked
        elif wedge.identity != blocked.identity:
            wedged_every_attempt = False

    # The wedged remedy tells the caller to delete the lockfile, so it may only be
    # given while the wedged file is the one still standing at the lock path. A
    # lock that moved on belongs to whoever published it — a run that may well be
    # alive — and contention is what this run really met.
    if wedged_every_attempt and wedge is not None and still_names_file(lock_path, wedge.identity):
        raise _wedged_takeover_error(lock_path, wedge)

    message = f"Lock at {lock_path} was claimed by another process on every attempt."
    hint = f"{_LIVE_HOLDER_HINT} If no gymrat process is running, delete {lock_path}."
    raise GymratError(message, hint=hint)
